qiskit_experiments/data_processing/data_action.py

```diff
@@ -0,0 +1,74 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Defines the steps that can be used to analyse data."""
+
+from abc import ABCMeta, abstractmethod
+from typing import Any
+
+
+class DataAction(metaclass=ABCMeta):
+    """
+    Abstract action which is a single action done on measured data to process it.
+    Each subclass of DataAction must define the type of data that it accepts as input
+    using decorators.
+    """
+
+    def __init__(self):
+        """Create new data analysis routine."""
+        self._child = None
+
+    @abstractmethod
+    def _process(self, datum: Any) -> Any:
+        """
+        Applies the data processing step to the data.
+
+        Args:
+            datum: A single item of data data which will be processed.
+
+        Returns:
+            processed data: The data that has been processed.
+        """
+
+    @abstractmethod
+    def _check_data_format(self, datum: Any) -> Any:
```

Comment:
The name is a little confusion since it is really formatting the input data, and doing validation as part of that.

Maybe this function should be:
```python
def _format_data(self, datum: Any, validate: bool = True) -> Any:
```
So that validation can be made optional (you might want to disable it for performance when you know your data is going to be valid).

To make use optional validation we would need something like adding `validate` as a kwarg to `__init__` like:
```python
def __init__(self, validate=True):
    self._validate = validate
```
and then in call:
```python
def __call__(self, data):
    self._process(self._format_data(data, validate=self._validate))
```

Reply:
See here: 4899bfa